core/rpc/src/state/state_light.rs

```diff
@@ -147,26 +207,15 @@ impl<Block, F, B, E, RA> StateBackend<B, E, Block, RA> for LightState<Block, F,
 		block: Option<Block::Hash>,
 		key: StorageKey,
 	) -> FutureResult<Option<StorageData>> {
-		let fetcher = self.fetcher.clone();
-		let storage = self.resolve_header(block)
-			.then(move |result| match result {
-				Ok(header) => Either::Left(fetcher.remote_read(RemoteReadRequest {
-					block: header.hash(),
-					header,
-					keys: vec![key.0.clone()],
-					retry_count: Default::default(),
-				}).then(move |result| ready(result
-					.map(|mut data| data
-						.remove(&key.0)
-						.expect("successful result has entry for all keys; qed")
-						.map(StorageData)
-					)
-					.map_err(client_err)
-				))),
-				Err(error) => Either::Right(ready(Err(error))),
-			});
-
-		Box::new(storage.boxed().compat())
+		Box::new(storage(
+			&*self.remote_blockchain,
+			self.fetcher.clone(),
+			self.block_or_best(block),
+			vec![key.0.clone()],
+		).boxed().compat().map(move |mut values| values
+			.remove(&key)
+			.expect("successful request has entries for all requested keys; qed")
```

Comment:
Is that really verified somewhere? I wonder if a malicious full node sends us an empty response and we panic here.

Reply:
Yes - there's so called `FetchChecker` which checks every response from remote node on light client. For remote read it tries to read every requested value from the response && fails if there's something wrong with the response. Otherwise - value is inserted into `HashMap`. This happens [here](https://github.com/paritytech/substrate/blob/master/core/state-machine/src/lib.rs#L624). So if everything goes that way, this expect should be valid